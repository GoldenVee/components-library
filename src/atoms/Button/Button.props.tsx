import { ComponentPropsWithRef } from 'react';

export interface ButtonProps extends ComponentPropsWithRef<'button'> {
  name?: string | '';
  size?: 'xs' | 'small' | 'medium' | 'ml' | 'large' | 'xl';
  variant?:
    | 'primary'
    | 'secondary'
    | 'tertiary'
    | 'quarterary'
    | 'transparent'
    | 'confirm'
    | 'destroy'
    | 'gradient';
  border?: 'none' | 'small' | 'medium';
  borderRadius?: 'none' | 'small' | 'medium' | 'large' | 'round';
  boxShadow?: 'none' | 'small' | 'large' | 'inset';
  fullWidth?: boolean;
  disabled?: boolean;
  hasIcon?: boolean;
  iconPosition?: 'leading' | 'trailing';
  mediaSrc?: string | '';
  mediaAlt?: string | '';
  indicator?: string;
  children?: React.ReactNode;
}
