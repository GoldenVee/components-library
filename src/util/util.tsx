import React from 'react';
import { ThemeProvider } from 'styled-components';
import { themes } from '../themes';

export const ThemeWrapped = (Story: React.ReactElement): React.ReactElement => {
  return (
    <ThemeProvider theme={themes['light'] || themes['dark']}>
      {Story}
    </ThemeProvider>
  );
};
