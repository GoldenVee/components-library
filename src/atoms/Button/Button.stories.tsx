import React from 'react';
import { Meta, Story } from '@storybook/react';
import { Button } from './Button';
import styled from 'styled-components';
import { ButtonProps } from './Button.props';
import starIcon from '../../assets/images/star-icon.png';

const meta: Meta = {
  title: 'atoms/Button',
  component: Button,
  argTypes: {
    name: {
      description:
        'Label of button, also used as aria text for screen readers.',
      control: {
        type: 'text',
      },
    },
    size: {
      description: 'Size of overall button.',
      control: {
        type: 'radio',
        options: ['xs', 'small', 'medium', 'ml', 'large', 'xl'],
      },
    },
    variant: {
      description:
        'Color palette of button. Determines background, font, border, and hover colors.',
      control: {
        type: 'radio',
        options: [
          'primary',
          'secondary',
          'tertiary',
          'quarterary',
          'transparent',
          'gradient',
        ],
      },
    },
    border: {
      description: 'Toggle for border.',
      control: {
        type: 'radio',
        options: ['none', 'small', 'medium'],
      },
    },
    borderRadius: {
      description: 'Determine the intensity of the border radius.',
      control: {
        type: 'radio',
        options: ['small', 'medium', 'large', 'round', 'none'],
      },
    },
    boxShadow: {
      description: 'Determine the size of the box shadow.',
      control: {
        type: 'radio',
        options: ['small', 'large', 'inset', 'none'],
      },
    },
    fullWidth: {
      description: 'Allow button to stretch the full width of the container.',
      control: {
        type: 'boolean',
      },
    },
    hasIcon: {
      description:
        'Allows for an icon or media src and positions it trailing or leading.',
      control: {
        type: 'boolean',
      },
    },
    iconPosition: {
      description: 'Choose between an icon in front of the text or behind it.',
      control: {
        type: 'radio',
        options: ['leading', 'trailing'],
      },
    },
    mediaSrc: {
      description: 'Provide the source of the image or icon you want to use.',
      control: {
        type: 'text',
      },
    },
    mediaAlt: {
      description:
        'Provide the alt text for the image or icon you want to use.',
      control: {
        type: 'text',
      },
    },
    disabled: {
      description: 'Toggles disabled state of onClick.',
      control: {
        type: 'boolean',
      },
    },
    onClick: { action: 'clicked' },
  },
  parameters: {
    controls: { expanded: true },
  },
};

export default meta;

const ExampleContainer = styled.div`
  display: flex;
  flex-wrap: wrap;
`;

const Template: Story<ButtonProps> = (args) => {
  return (
    <ExampleContainer>
      <Button {...args}></Button>
    </ExampleContainer>
  );
};

const SizesTemplate: Story<ButtonProps> = (args) => (
  <ExampleContainer>
    <Button size="xs" {...args} />
    <Button size="small" {...args} />
    <Button size="medium" {...args} />
    <Button size="ml" {...args} />
    <Button size="large" {...args} />
    <Button size="xl" {...args} />
  </ExampleContainer>
);

const VariantsTemplate: Story<ButtonProps> = (args) => (
  <ExampleContainer>
    <Button name="Primary" variant="primary" {...args} />
    <Button name="Secondary" variant="secondary" {...args} />
    <Button name="Tertiary" variant="tertiary" {...args} />
    <Button name="Quarterary" variant="quarterary" {...args} />
    <Button name="Transparent" variant="transparent" {...args} />
    <Button name="Gradient" variant="gradient" {...args} />
  </ExampleContainer>
);

const BordersTemplate: Story<ButtonProps> = (args) => (
  <ExampleContainer>
    <Button name="Small radius" borderRadius="small" {...args} />
    <Button name="Medium radius" borderRadius="medium" {...args} />
    <Button name="Large radius" borderRadius="large" {...args} />
    <Button name="Round radius" borderRadius="round" {...args} />
    <Button name="No radius" borderRadius="none" {...args} />
    <Button
      name="Medium border"
      border="medium"
      borderRadius="medium"
      {...args}
    />
    <Button
      name="Small border"
      border="small"
      borderRadius="medium"
      {...args}
    />
  </ExampleContainer>
);

const BoxShadowTemplate: Story<ButtonProps> = (args) => (
  <ExampleContainer>
    <Button name="Small shadow" boxShadow="small" {...args} />
    <Button name="Large shadow" boxShadow="large" {...args} />
    <Button name="Inset shadow" boxShadow="inset" {...args} />
    <Button name="No box shadow" boxShadow="none" {...args} />
  </ExampleContainer>
);

const FullWidthTemplate: Story<ButtonProps> = (args) => (
  <ExampleContainer>
    <Button
      name="Full container width button"
      border="small"
      borderRadius="medium"
      fullWidth={'true'}
      {...args}
    />
    <Button name="Normal" border="medium" borderRadius="medium" {...args} />
  </ExampleContainer>
);
const IconButtonTemplate: Story<ButtonProps> = (args) => (
  <ExampleContainer>
    <Button
      name="Leading Icon"
      iconPosition="leading"
      mediaSrc={starIcon}
      {...args}
    />
    <Button
      name="Trailing Icon"
      iconPosition="trailing"
      mediaSrc={starIcon}
      {...args}
    />
  </ExampleContainer>
);

const DisabledTemplate: Story<ButtonProps> = (args) => (
  <ExampleContainer>
    <Button name="Normal" {...args} />
    <Button name="Disabled" disabled {...args} />
  </ExampleContainer>
);

const ChildrenTemplate: Story<ButtonProps> = (args) => (
  <ExampleContainer>
    <Button name="Normal" {...args}>
      <div className="pointer-events-auto flex items-center justify-between gap-x-6 bg-gray-600 px-6 py-2.5 sm:rounded-xl sm:py-3 sm:pl-4 sm:pr-3.5">
        <p className="text-sm leading-6 text-white">
          <a href="#">
            <strong className="font-semibold">Lorem ipsum</strong>
            <svg
              viewBox="0 0 2 2"
              className="mx-2 inline h-0.5 w-0.5 fill-current"
              aria-hidden="true"
            >
              <circle cx={1} cy={1} r={1} />
            </svg>
            Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
            eiusmod tempor incididunt ut labore et dolore magna aliqua.
            <span aria-hidden="true">&rarr;</span>
          </a>
        </p>
      </div>
    </Button>
  </ExampleContainer>
);

// By passing using the Args format for exported stories, you can control the props for a component for reuse in a test
// https://storybook.js.org/docs/react/workflows/unit-testing
export const Default = Template.bind({});
Default.args = {
  name: 'Click me',
  variant: 'primary',
  size: 'large',
  border: 'small',
  borderRadius: 'medium',
  boxShadow: 'small',
  fullWidth: false,
};

export const Sizes = SizesTemplate.bind({});
Sizes.args = {
  name: 'Click me',
  variant: 'primary',
};

export const Variants = VariantsTemplate.bind({});
Variants.args = {
  size: 'large',
};

export const Borders = BordersTemplate.bind({});
Borders.args = {
  size: 'large',
  variant: 'primary',
};

export const BoxShadow = BoxShadowTemplate.bind({});
BoxShadow.args = {
  variant: 'tertiary',
  size: 'large',
  border: 'none',
  borderRadius: 'medium',
};

export const FullWidth = FullWidthTemplate.bind({});
FullWidth.args = {
  variant: 'primary',
  border: 'small',
};

export const IconButton = IconButtonTemplate.bind({});
IconButton.args = {
  variant: 'primary',
  size: 'large',
  border: 'small',
  borderRadius: 'medium',
  boxShadow: 'small',
  fullWidth: false,
  hasIcon: true,
};

export const DisabledButton = DisabledTemplate.bind({});
DisabledButton.args = {
  variant: 'primary',
  size: 'large',
  border: 'small',
  borderRadius: 'medium',
  boxShadow: 'small',
  fullWidth: false,
  hasIcon: false,
};

export const ChildrenButton = ChildrenTemplate.bind({});
ChildrenButton.args = {};
