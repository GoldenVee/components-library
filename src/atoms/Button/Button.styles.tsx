import styled from 'styled-components';
import tw, { TwStyle } from 'twin.macro';
import { ThemeProps, styledThemeProps } from '../../themes';
import { ButtonProps } from './Button.props';

export const StyledButton = styled.button<ButtonProps>(
  ({ variant, fullWidth, disabled, theme }) => [
    ({ variant = 'primary' }) => Variants[variant],
    ({ size = 'medium' }) => Sizes[size],
    ({ border = 'none' }) => Border[border],
    ({ borderRadius = 'medium' }) => BorderRadius[borderRadius],
    ({ boxShadow = 'small' }) => BoxShadow[boxShadow],
    fullWidth && tw`flex w-full justify-center`,
    disabled && theme.colors.disabled,
    disabled &&
      variant === 'transparent' &&
      tw`disabled:opacity-75 bg-slate-200`,
    tw`flex flex-row justify-center items-center`,
  ],
);

export const LabelContainer = styled.div<ButtonProps>(({ hasIcon }) => [
  tw`flex w-auto`,
  hasIcon && tw`mx-2`,
]);

export const Sizes: TwStyle = {
  xs: tw`px-2 py-1.5 m-2 text-xs h-min`,
  small: tw`px-2.5 py-1.5 m-2 text-sm h-min`,
  medium: tw`px-3.5 py-2 m-2 text-sm h-min`,
  ml: tw`px-4 py-3 m-2 text-sm h-min`,
  large: tw`px-5 py-3.5 m-2 text-sm h-min`,
  xl: tw`px-6 py-4 m-2 text-base h-min`,
};

export const Variants: styledThemeProps = {
  primary: ({ theme }: { theme: ThemeProps }) => theme.colors.primary,
  secondary: ({ theme }: { theme: ThemeProps }) => theme.colors.secondary,
  tertiary: ({ theme }: { theme: ThemeProps }) => theme.colors.tertiary,
  quarterary: ({ theme }: { theme: ThemeProps }) => theme.colors.quaternary,
  transparent: ({ theme }: { theme: ThemeProps }) => theme.colors.transparent,
  destroy: ({ theme }: { theme: ThemeProps }) => theme.colors.destroy,
  confirm: ({ theme }: { theme: ThemeProps }) => theme.colors.confirm,
  gradient: ({ theme }: { theme: ThemeProps }) => theme.colors.gradient,
};

export const Border: styledThemeProps = {
  none: ({ theme }: { theme: ThemeProps }) => theme.border.noBorder,
  small: ({ theme }: { theme: ThemeProps }) => theme.border.small,
  medium: ({ theme }: { theme: ThemeProps }) => theme.border.medium,
};

export const BorderRadius: styledThemeProps = {
  none: ({ theme }: { theme: ThemeProps }) => theme.borderRadius.none,
  small: ({ theme }: { theme: ThemeProps }) => theme.borderRadius.small,
  medium: ({ theme }: { theme: ThemeProps }) => theme.borderRadius.medium,
  large: ({ theme }: { theme: ThemeProps }) => theme.borderRadius.large,
  round: ({ theme }: { theme: ThemeProps }) => theme.borderRadius.round,
};

export const BoxShadow: styledThemeProps = {
  none: ({ theme }: { theme: ThemeProps }) => theme.boxShadow.none,
  small: ({ theme }: { theme: ThemeProps }) => theme.boxShadow.small,
  large: ({ theme }: { theme: ThemeProps }) => theme.boxShadow.large,
  inset: ({ theme }: { theme: ThemeProps }) => theme.boxShadow.inset,
};
