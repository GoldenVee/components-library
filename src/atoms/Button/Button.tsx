import React, { forwardRef, ComponentPropsWithoutRef } from 'react';
import { StyledButton } from './Button.styles';
import { ButtonProps } from './Button.props';
import { BeakerIcon } from '@heroicons/react/24/outline';

export type Ref = HTMLButtonElement;

const renderMedia = ({ mediaSrc, mediaAlt }: ButtonProps): JSX.Element => {
  return (
    <BeakerIcon className="h-5 w-5 mx-1 text-yellow-500 mix-blend-difference" />
  );
};

const handleClick = ({ disabled, onClick }: ButtonProps) => {
  return disabled ? undefined : onClick;
};

export const Button = forwardRef<Ref, ComponentPropsWithoutRef<'button'>>(
  (
    {
      name,
      size = 'medium',
      variant = 'primary',
      border = 'small',
      borderRadius = 'medium',
      boxShadow = 'small',
      fullWidth = false,
      hasIcon = false,
      iconPosition = 'leading',
      mediaSrc = '',
      mediaAlt = '',
      disabled = false,
      children,
      onClick,
      ...props
    }: ButtonProps,
    ref,
  ): JSX.Element => {
    return children ? (
      <button onClick={handleClick({ disabled, onClick })}>{children}</button>
    ) : (
      <StyledButton
        size={size}
        variant={variant}
        border={border}
        borderRadius={borderRadius}
        boxShadow={boxShadow}
        fullWidth={fullWidth}
        hasIcon={hasIcon}
        iconPosition={iconPosition}
        mediaSrc={mediaSrc}
        mediaAlt={mediaAlt}
        disabled={disabled}
        onClick={handleClick({ disabled, onClick })}
        ref={ref}
        {...props}
        aria-label={name}
      >
        {hasIcon &&
          iconPosition === 'leading' &&
          renderMedia({ mediaSrc, mediaAlt })}
        {name}
        {hasIcon &&
          iconPosition === 'trailing' &&
          renderMedia({ mediaSrc, mediaAlt })}
      </StyledButton>
    );
  },
);

Button.displayName = 'Button';
