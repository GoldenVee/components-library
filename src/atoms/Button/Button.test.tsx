import React from 'react';
import { render } from '@testing-library/react';
import { Default } from './Button.stories';
import { ThemeWrapped } from '../../util/util';

describe('Button', () => {
  it('renders as expected', () => {
    const { container } = render(ThemeWrapped(<Default name="test" />));
    expect(container).toMatchSnapshot();
  });
});
