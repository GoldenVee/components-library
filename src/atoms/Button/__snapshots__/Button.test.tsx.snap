// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders as expected 1`] = `
<div>
  <div
    class="Button__CardExample-sc-1aocd7z-0 fgjxid"
  >
    <button
      class="Buttonstyles__sharedButtonStyling-sc-iaqf7v-0 Buttonstyles__StyledButtonOne-sc-iaqf7v-1 lczdqw ebxeau"
      type="button"
    >
      test
    </button>
    <button
      class="Buttonstyles__sharedButtonStyling-sc-iaqf7v-0 Buttonstyles__StyledButtonTwo-sc-iaqf7v-2 lczdqw xXUiV"
      type="button"
    >
      test
    </button>
    <button
      class="Buttonstyles__sharedButtonStyling-sc-iaqf7v-0 Buttonstyles__StyledButtonThree-sc-iaqf7v-3 lczdqw igOyGn"
      type="button"
    >
      test
    </button>
    <p
      class="Button__TestText-sc-1aocd7z-1 kwKoEh"
    >
      Text test
    </p>
    <p
      class="Button__TestSubText-sc-1aocd7z-2 cyjOwK"
    >
      Subtext test
    </p>
    <a
      class="Button__TextLink-sc-1aocd7z-3 bxqyim"
      href="https://whitelabelco.com/blog"
    >
      Link test
    </a>
    <button
      class="Buttonstyles__TwButton-sc-iaqf7v-4 bWciPT"
    >
      This is a twin.macro-styled button
    </button>
  </div>
</div>
`;
